### PR TITLE
simple node http server for webpack dev

### PR DESCRIPTION
**Pull request by yishengjiang99**
rebased to

### CODE REVIEW

File: packages/xarc-app-dev/src/lib/dev-admin/dev-http.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
+/* eslint-disable @typescript-eslint/no-var-requires */
+export {};
```
Comment (jchip) [inline packages/xarc-app-dev/src/lib/dev-admin/dev-http.ts:2]: no longer needed

File: packages/xarc-app-dev/src/lib/dev-admin/dev-server.ts [hunk 2]
```diff
@@ -11,6 +12,7 @@ const electrodeServer = optionalRequire("electrode-server");
 const Hapi = optionalRequire("@hapi/hapi");
 const Koa = optionalRequire("koa");
 const express = optionalRequire("express");
+const http = optionalRequire("http");
```
Comment (jchip) [inline packages/xarc-app-dev/src/lib/dev-admin/dev-server.ts:15]: `http` is a built-in module, no need to `optionalRequire` it.
also, use `import`.

File: packages/xarc-app-dev/src/lib/dev-admin/dev-server.ts [hunk 3]
```diff
@@ -20,7 +22,21 @@ if (process.env.WEBPACK_DEV === undefined) {
   process.env.WEBPACK_DEV = "true";
 }
 
-if (fastifyServer) {
+if (http) {
```
Comment (jchip) [inline packages/xarc-app-dev/src/lib/dev-admin/dev-server.ts:25]: this would always be true.

**Comment by jchip** [comment]
CI failed,  run `xrun check` locally to make sure it pass

**Review by jcaowml — APPROVED** [review]

File: packages/xarc-app-dev/src/lib/dev-admin/dev-http.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
+/* eslint-disable @typescript-eslint/no-var-requires */
+import { createReadStream } from "fs";
+import { createServer, IncomingMessage, RequestListener, ServerResponse } from "http";
+import * as Url from "url";
+import { resolve } from "path";
+const Middleware = require("./middleware");
```
Comment (jcaowml) [inline packages/xarc-app-dev/src/lib/dev-admin/dev-http.ts:6]: I think middleware is typed, so you can import it normally

File: packages/xarc-app-dev/src/lib/dev-admin/dev-http.ts [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+      skip: () => Promise.resolve(),
+      replyHtml: html => {
+        res
+          .writeHead(200, {
+            "Content-Type": "text/html"
+          })
+          .end(`<!DOCTYPE html>${html}`);
+      },
+      replyError: err => {
+        res.writeHead(500, err);
```
Comment (jcaowml) [inline packages/xarc-app-dev/src/lib/dev-admin/dev-http.ts:61]: Is res.end() needed?

File: packages/xarc-app-dev/src/lib/dev-admin/dev-http.ts [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+  const server = createServer(webpackDevHttpPlugin);
+
+  return {
+    start: () => server.listen(port, host),
+    webpackDevHttpPlugin,
+    addServerEventListener: (event: HttpServerEvent, cb) => {
+      server.addListener(event.toString(), cb);
+    },
+    stop: () => {
+      server.close(function() {
```
Comment (jcaowml) [inline packages/xarc-app-dev/src/lib/dev-admin/dev-http.ts:83]: Since this is asynchronous, it's a good idea to make this return a promise?

**Pull request merged**